streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -789,6 +790,14 @@ private static ScheduledExecutorService maybeCreateRocksDBMetricsRecordingServic
         return null;
     }
 
+    private static void maybeWarnAboutCodeInRocksDBConfigSetter(final Logger log,
+                                                                final StreamsConfig config) {
+        if (config.getClass(StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG) != null) {
+            log.warn("Kafka Streams cannot guarantee backwards compatibility of code in the RocksDB config setter "
+                + "because RocksDB may remove methods without deprecating them.");
```

Comment:
Please feel free to propose a better phrasing.

I tried to keep it general, because backwards-incombatible RocksDBConfigSetters can also happen in future after we upgrade to a newer version when RocksDB decides again to remove a method without deprecation.

Reply:
I would be more concrete:
```
RocksDB version will be bumped to version 6+ via KAFKA-8897 in a future release.
If you use `org.rocksdbOptions#WhatEverMethodWasRemoved`, you will need to rewrite your code after KAFKA-8897 is resolved because those methods were removed from `org.rocksdbOptions` class without a deprecation period.
```

Or something like this.

Reply:
I am fine with being more specific and I find the pointer to the Jira ticket a good idea. However,
a) I would not list all removed methods because I looked into the changes and it is not only `CompactionOptionsFIFO#setTtl()` that was removed but also other methods. It seems to me a lot of work for little result to research all methods removed without deprecation period. Also because we do not yet know to which version we will increase RocksDB. Additionally, it is very easy to make a mistake here.
b) IMO, we need to clearly state in the docs (and maybe in the logs) that we do not guarantee backwards compatibility of the code in the config setter because we do not control RocksDB's API. Otherwise, we will run into the same issue again and again.

Reply:
What other methods were removed? From what I saw there was only the one relevant method, ie we only need to warn users about removed methods that are
a) public
b) in their Java API
c) visible/relevant to the store Options/configuration

I agree we should not guarantee backwards compatibility of RocksDB APIs, but I wouldn't put that in a log message (just docs). That said, we are still somewhat responsible for them as we have control over when to upgrade Rocks, and to which version. We shouldn't suggest we will just blindly upgrade and that is it up to users to go through the rocksdb docs and figure out if any relevant methods were removed, anytime they want to upgrade.

Reply:
>  It seems to me a lot of work for little result to research all methods removed without deprecation period.
Exactly -- we shouldn't push this to every user on every upgrade. It's annoying, so, we should just do it exactly once, and only before considering an upgrade.

Reply:
The following code will not compile on a RocksDB version 6+
``` 
Env defaultEnv = Env.getDefault();
defaultEnv.setBackgroundThreads(2, Env.FLUSH_POOL);
options.setEnv(defaultEnv);
```
because the signature of `setBackgroundThreads()` changed from

```
public Env setBackgroundThreads(final int num, final int poolID)
```
to 
```
public Env setBackgroundThreads(final int num, final Priority priority)
```
without deprecation period and without keeping the old method. 

Similar holds for `getThreadPoolQueueLen()`. I just looked into the `Env` class. I haven't checked `BuiltinComparator`, `MergeOperator` and all the other classes that are passed as parameters to the methods in class `Options`.

I still think that is too much work for little outcome to check an API for which we are not responsible. We just provide a hook to use this API.

If we clearly state an upgrade of RocksDB in the release notes and remind the users that there is a possibility that their config setter will not compile, then we do not suggest that we blindly upgrade RocksDB.